### PR TITLE
Setup ORM with alembic for migrations

### PR DESCRIPTION
**Pull request by null-domain**

### CODE REVIEW

File: bannou/client.py [hunk 1]
```diff
@@ -1,9 +1,12 @@
 import pathlib
 
+
+from databases import Database as DatabaseType
```
Comment (davfsa) [inline bannou/client.py:4]: As a coding practice, should we keep in-line with hikari and import the modules and then refer to them? I find it nicer to then pinpoint where classes are coming from and being defined

File: bannou/database/context.py [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+
+from bannou import bot_settings
+
+_database = Database(bot_settings.postgres.build_url())
+_metadata = sqlalchemy.MetaData()
+
+
+class BaseMeta(ModelMeta):
+    database = _database
+    metadata = _metadata
```
Comment (davfsa) [inline bannou/database/context.py:13]:
```suggestion
class BaseMeta(ModelMeta):
    database = Database(bot_settings.postgres.build_url())
    metadata = sqlalchemy.MetaData()
```

File: bannou/database/migrations/env.py [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+
+config.set_main_option("sqlalchemy.url", str(bannou.database.context.BaseMeta.database.url))
+
+
+def run_migrations_offline() -> None:
+    """Run migrations in 'offline' mode.
+
+    This configures the context with just a URL
+    and not an Engine, though an Engine is acceptable
+    here as well.  By skipping the Engine creation
```
Comment (davfsa) [inline bannou/database/migrations/env.py:40]:
```suggestion
    here as well. By skipping the Engine creation
```

File: bannou/database/migrations/env.py [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+    async with connectable.connect() as connection:
+        await connection.run_sync(do_run_migrations)
+
+    await connectable.dispose()
+
+
+if context.is_offline_mode():
+    run_migrations_offline()
+else:
+    asyncio.run(run_migrations_online())
```
Comment (davfsa) [inline bannou/database/migrations/env.py:91]: If this is going to be a script, wouldnt it be better to keep it outside of the bot source code and into a `scripts` folder?

File: bannou/database/models/__init__.py [hunk 5]
```diff
@@ -0,0 +1,3 @@
+from ..context import BaseMeta
+
+from .tags import Tag
```
Comment (davfsa) [inline bannou/database/models/__init__.py:3]:
```suggestion
from bannou.database.context import BaseMeta

from bannou.database.models.tags import Tag
```

Relative imports can be really hard to follow and maintain

File: bannou/database/models/tags.py [hunk 6]
```diff
@@ -0,0 +1,25 @@
+from __future__ import annotations
+
+from typing import TYPE_CHECKING
+
+import ormar
+import sqlalchemy
+
+from . import BaseMeta
```
Comment (davfsa) [inline bannou/database/models/tags.py:8]: Same here

File: pyproject.toml [hunk 7]
```diff
@@ -8,3 +8,7 @@ profile = "black"
 [tool.ruff]
 line-length = 120
 target-version = 'py311'
+
+[tool.ruff.per-file-ignores]
+"bannou/database/migrations/env.py" = ["F401"]
+"bannou/database/models/__init__.py" = ["F401"]
```
Comment (davfsa) [inline pyproject.toml:14]: Missing new line

File: pyproject.toml [hunk 8]
```diff
@@ -8,3 +8,7 @@ profile = "black"
 [tool.ruff]
 line-length = 120
 target-version = 'py311'
+
+[tool.ruff.per-file-ignores]
+"bannou/database/migrations/env.py" = ["F401"]
```
Comment (davfsa) [inline pyproject.toml:13]: Can we try to document these codes right above with a meaningful description on why they are ignored?

File: bannou/__init__.py [hunk 9]
```diff
@@ -0,0 +1,3 @@
+from bannou import settings
+
+bot_settings = settings.BotSettings("config.yaml")
```
Comment (davfsa) [inline bannou/__init__.py:3]: Dont like this change at all. This would mean that settings are read off at import time, which should only be happening when needed. Init files should not contain any logic unless strictly necessary. Not only that, this also assumes that the module is being imported from the path that config.yaml is in.

For the single place where this is being "useful", we can just add an init to the class and pass the settings.

File: bannou/database/migrations/env.py [hunk 10]
```diff
@@ -0,0 +1,91 @@
+import asyncio
+from logging.config import fileConfig
+
+from alembic import context
+from sqlalchemy import engine_from_config, pool
+from sqlalchemy.engine import Connection
+from sqlalchemy.ext.asyncio import AsyncEngine
+
+import bannou.database.context
```
Comment (davfsa) [inline bannou/database/migrations/env.py:9]: To differentiate alembic context and bannou context, we can use `as alembic_context` and `as bannou_context` respectively

**Review by davfsa — CHANGES_REQUESTED** [review]

File: bannou/client.py (hunk 1 above)
Comment (null-domain) [reply bannou/client.py:4]: Can do

File: bannou/database/context.py (hunk 2 above)
Comment (null-domain) [reply bannou/database/context.py:13]: I had it like this before but changed it to what it is now - can't actually remember _why_ I did that. Can always change it back if it breaks but I don't think it should, probably was just me being an idiot

File: bannou/database/migrations/env.py (hunk 4 above)
Comment (null-domain) [reply bannou/database/migrations/env.py:91]: Unfortunately the directory structure of `bannou/database/migrations` is set up and managed by alembic, I'm not sure if we can change the location of this.

File: bannou/database/models/__init__.py (hunk 5 above)
Comment (null-domain) [reply bannou/database/models/__init__.py:3]: Gotcha, will keep to this practice

File: bannou/__init__.py (hunk 9 above)
Comment (null-domain) [reply bannou/__init__.py:3]: Not entirely sure how I would go about doing this 'cleanly'. Will mention it tomorrow

File: bannou/database/migrations/env.py (hunk 4 above)
Comment (null-domain) [reply bannou/database/migrations/env.py:91]: Actually I think we can. Will play around and see what I can achieve

File: bannou/client.py [hunk 11]
```diff
@@ -1,9 +1,12 @@
 import pathlib
 
+
+import databases
 import hikari
 import tanjun
 
 from bannou.settings import BotSettings
+from bannou.database.context import BaseMeta
```
Comment (davfsa) [inline bannou/client.py:9]:
```suggestion
from bannou import settings
from bannou.database import context
```

File: bannou/client.py [hunk 12]
```diff
@@ -1,9 +1,12 @@
 import pathlib
 
+
+import databases
```
Comment (davfsa) [inline bannou/client.py:4]: Out of curiosity, where does this import come from?

File: bannou/database/migrations/env.py [hunk 13]
```diff
@@ -0,0 +1,92 @@
...
+from sqlalchemy import engine_from_config, pool
+from sqlalchemy.engine import Connection
+from sqlalchemy.ext.asyncio import AsyncEngine
+
+import bannou.database.context as bannou_context
+
+# we import the models even though they aren't used directly
+# otherwise alembic will not detect them
+# ignored via F401 in per-file-ignores in pyproject.toml
+from bannou.database import models
```
Comment (davfsa) [inline bannou/database/migrations/env.py:14]: If F401 is only for here, cant we just ignore it in-line?

File: bannou/database/context.py [hunk 14]
```diff
@@ -0,0 +1,11 @@
+import sqlalchemy
+from databases import Database
+from ormar import ModelMeta
+
+from bannou import settings
+
+
+# TODO: Use already-instantiated bot settings (see PR 9)
+class BaseMeta(ModelMeta):
+    database = Database(settings.BotSettings("config.yaml").postgres.build_url())
```
Comment (davfsa) [inline bannou/database/context.py:10]: Ngl, kinda hate this even more.

Instead of having it in the init, what about having settings.py export the loaded class already?

Will most probably make it tricky in the future to test, if we do get around to setting that up, but should be doable. This is loading the config file anyways.

File: bannou/client.py (hunk 11 above)
Comment (null-domain) [reply bannou/client.py:9]: The change on line 8 would affect the parameter name for `build_bot(settings: ...)`.
`from bannou.settings import BotSettings` was how you originally did it in 0778386.

File: bannou/client.py (hunk 12 above)
Comment (null-domain) [reply bannou/client.py:4]: This import is for [encode/databases](https://github.com/encode/databases). The only place we use this import or its classes directly is for connecting and closing connections to the database, and for the BaseMeta class for DB models.

**Review by davfsa — APPROVED** [review]
